synapse/storage/databases/main/push_rule.py

```diff
@@ -464,9 +465,17 @@ def _add_push_rule_highest_priority_txn(
         conditions_json: str,
         actions_json: str,
     ) -> None:
-        # Lock the table since otherwise we'll have annoying races between the
-        # SELECT here and the UPSERT below.
-        self.database_engine.lock_table(txn, "push_rules")
+        if isinstance(self.database_engine, PostgresEngine):
+            # Postgres doesn't do FOR SHARE on aggregate functions, so select the rows first
+            # then re-select the count/max below.
+            sql = """
+                SELECT * FROM push_rules
+                WHERE user_name = ? and priority_class = ?
+                FOR SHARE
+            """
```

Comment:
Might be worth doing a `txn.execute` for this SQL? :laughing:

Reply:
But also, you probably want to do this as part of the `COUNT(*)` query below?

Reply:
Yep good spot - https://github.com/matrix-org/synapse/pull/16051/commits/2ec17da8985afb78ae619c72d7eaebccc38afbef

Reply:
Still missing a `tx.execute` here?